tagstudio/src/core/library.py

```diff
@@ -2324,3 +2172,287 @@ def sort_fields(self, entry_id: int, order: list[int]) -> None:
         entry.fields = sorted(
             entry.fields, key=lambda x: order.index(self.get_field_attr(x, "id"))
         )
+
+
+class Filter:
+    def __init__(self, lib: Library) -> None:
+        # self.lib = lib
+        self.ext_list = lib.ext_list
+        self.entries = lib.entries
+        self.is_exclude_list = lib.is_exclude_list
+        self.get_field_attr = lib.get_field_attr
+        self._tag_strings_to_id_map = lib._tag_strings_to_id_map
+        self.get_tag_cluster = lib.get_tag_cluster
+        self.library_dir = lib.library_dir
+        self.get_field_obj = lib.get_field_obj
+        self.missing_files = lib.missing_files
+        self.default_fields = lib.default_fields
+
+        def generate_field_id_to_name_map(self) -> dict[int, str]:
+            map: dict = {}
+            for field in self.default_fields:
+                map[field.get("id")] = field.get("name").lower()
+            return map
+
+        self._field_id_to_name_map: dict = generate_field_id_to_name_map(self)
+
+    def filter_results(
+        self, query: str, split_query: list[dict], search_mode: SearchMode
+    ) -> list[tuple[ItemType, int]]:
+        # start_time = time.time()
+        query = query.strip().lower()
+
+        pre_results: list[list[tuple[ItemType, int]]] = []
+
+        for query_part in split_query:
+            all_tag_terms: list[str] = []
+            query_words = query_part.get("unbound")
+
+            if isinstance(query_words, list):
+                all_tag_terms.extend(self.preprocess_tag_terms(query_words))
+
+            filtered_entries: list = []
+
+            # Iterate over all Entries =============================================================
+            for entry in self.entries:
+                entry_tuple: tuple | None = None
+                allowed_ext: bool = entry.filename.suffix not in self.ext_list
+                if allowed_ext != self.is_exclude_list:
+                    continue
+                (entry_tags, entry_authors) = self.populate_tags(entry)
```

Comment:
the tags and authors (among others) are extracted into `entry_fields` a few lines below via `self.remap_fields(entry)`, so this feels redundant.

Reply:
This is needed for special queries like 'no authors' to work, since there are multiple fields with types `tag_box` and names related to `Author`. Although I could create loops to check every field if it is a `tag_box` or not, I think current solution is preffered, let me know what you think

Reply:
I'd assume the negative syntax needs to work for every field, otherwise that would be a bit non-intuitive that it works only for tags and authors 🤔

Reply:
I see. And I dont know. This, together with the difficult way how the negative queries like 'no artist' are parsed, makes me really think that the syntax `-artist` would be more straightforward and easier to work with. I'm here just to review code for fun, but I dont make the product decision.